### DISCUSSION TITLE
Pre-loading with filter

### DISCUSSION THREAD

**Discussion by lukas-krecan** [Netflix/dgs-framework#1455] [Q&A]
Let's say I have the following schema

```
type Query {
  shows: [Show]
}

type Show {
  # The showId may or may not be there, depending on the scenario.
  showId: ID
  title: String
  reviews(filter: MyFilter): [Review]
}

type Review {
  starRating: Int
  text: String
}
```

I want to use [pre-loading](https://netflix.github.io/dgs/advanced/context-passing/#pre-loading) to fetch reviews in one DB query with shows as I can easily join it in the DB. Moreover implementing it effectively using DataLoaders is far from straightforward. The question is how to access the review filter `MyFilter` from the shows fetcher?

**Comment by lukas-krecan** [comment]
The closest we were able to get was to somehow use the `ValuesResolver.literalToInternalValue` and then `DefaultInputObjectMapper` to deserialize it.

**Comment by srinivasankavitha** [comment]
This needs to be fixed in the schema, i.e. if there is an input that needs
to come the user, in this case the filter, you will need to update shows to
take filter as input argument, i.e.
```
type Query {
    shows(filter: Filter) : [Show]
}
```

If the filter is not provided by the user but will be available as part of
processing the shows query, you can pass it down to child data fetchers
using the local context. Check out an example here:
https://netflix.github.io/dgs/advanced/context-passing/#no-showid-use-local-context

You will need to get the local context from the data fetching environment
in this case.

Hope this helps.

**Reply by divyanshubhati** [reply]
Hi @srinivasankavitha,

Given the following schema and constraint:

```
type Query {
  shows(id: ID): Show!
}

type Show {
 
  showId: ID
  title: String
  author: String
  reviews(param: count): [Review]
}

type Review {
  review: String
}
```


There’s a strict requirement that if a user wants to fetch reviews for a show, they must provide a param value. Declaring the filter directly on the reviews field like `reviews(param: count): [Review] ` feels more natural than modifying the shows query to accept a filter argument—especially since users may not always want to fetch reviews.

The goal is to allow users to provide the param only when they want reviews, while still enabling a single service call to fetch both the show info and its reviews. To achieve this, I need access to the param value inside the Show data fetcher so it can be passed to the data loader.
What’s the best way to structure this so that:

The param is optional unless reviews are requested.
The param can be accessed within the Show resolver.
The entire data (show + reviews) can be fetched in one service call?
(NOTE: We do not want to declare the param in the shows query like` shows(id: ID, param: Count): Show!` )
Would like to know your thoughts on how to approach this.

**Reply by srinivasankavitha** [reply]
I think the look ahead technique might work better for you in this case. This approach lets you explore what is in the selection set from the data fetching environment: https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/
Once you get the field from the selection set, you should be able to look at the arguments, which will be `param` in your example.
